File: adapters/shared.ts
/**
 * Shared adapter utilities for Brain cross-platform plugin system.
 * Provides frontmatter parsing, file generation, and common transform helpers.
 */

import { join } from "path";

/** Read a text file using Bun.file(). */
export async function readTextFile(path: string): Promise<string> {
  return Bun.file(path).text();
}

// ─── Types ───────────────────────────────────────────────────────────────────

export interface TargetConfig {
  adapter?: string;
  outputDir?: string;
  agentFrontmatter?: Record<string, unknown>;
}

export interface BrainConfig {
  targets: Record<string, TargetConfig>;
  agents: Record<string, AgentToolConfig>;
  hooks?: Record<string, HookToolConfig>;
}

export interface AgentToolConfig {
  [tool: string]: AgentPlatformConfig | null;
}

export interface AgentPlatformConfig {
  model?: string;
  allowed_tools?: string[];
  memory?: string;
  color?: string;
  description?: string;
  argument_hint?: string;
  skills?: string[];
}

export interface HookToolConfig {
  [tool: string]: HookPlatformConfig | null;
}

export interface HookPlatformConfig {
  event: string;
  matcher?: string;
  timeout?: number;
}

export interface CanonicalAgent {
  /** Filename without extension (e.g., "architect") */
  name: string;
  /** Raw markdown body (everything after frontmatter) */
  body: string;
  /** Any frontmatter fields from the canonical file (minimal) */
  frontmatter: Record<string, unknown>;
}

export interface GeneratedFile {
  /** Relative path from the tool's config directory */
  relativePath: string;
  /** File content */
  content: string;
}

// ─── Frontmatter Parsing ─────────────────────────────────────────────────────

const FRONTMATTER_RE = /^---\n([\s\S]*?)\n---\n?([\s\S]*)$/;

/**
 * Parse a markdown file's YAML frontmatter and body.
 * Returns empty frontmatter if none found.
 */
export function parseFrontmatter(raw: string): {
  frontmatter: Record<string, unknown>;
  body: string;
} {
  const match = raw.match(FRONTMATTER_RE);
  if (!match) {
    return { frontmatter: {}, body: raw };
  }
  const [, yamlBlock, body] = match;
  const frontmatter = parseSimpleYaml(yamlBlock);
  return { frontmatter, body: body.trimStart() };
}

/**
 * Minimal YAML parser for agent frontmatter.
 * Handles: scalars, simple arrays (both inline [...] and block - item), nested objects (one level).
 * Does NOT handle complex nested structures, multi-line strings, or anchors.
 */
export function parseSimpleYaml(yaml: string): Record<string, unknown> {
  const result: Record<string, unknown> = {};
  const lines = yaml.split("\n");
  let currentKey = "";
  let currentArray: string[] | null = null;

  for (const line of lines) {
    // Skip empty lines and comments
    if (line.trim() === "" || line.trim().startsWith("#")) {
      if (currentArray && currentKey) {
        result[currentKey] = currentArray;
        currentArray = null;
        currentKey = "";
      }
      continue;
    }

    // Block array item (  - value)
    const arrayItemMatch = line.match(/^\s+-\s+(.+)$/);
    if (arrayItemMatch && currentArray) {
      currentArray.push(arrayItemMatch[1].trim());
      continue;
    }

    // Flush pending array
    if (currentArray && currentKey) {
      result[currentKey] = currentArray;
      currentArray = null;
    }

    // Key-value pair
    const kvMatch = line.match(/^(\w[\w-]*):\s*(.*)$/);
    if (!kvMatch) continue;

    const [, key, rawValue] = kvMatch;
    const value = rawValue.trim();

    // Inline array: [a, b, c]
    if (value.startsWith("[") && value.endsWith("]")) {
      const items = value
        .slice(1, -1)
        .split(",")
        .map((s) => s.trim())
        .filter(Boolean);
      result[key] = items;
      currentKey = "";
      continue;
    }

    // Empty value followed by array items
    if (value === "") {
      currentKey = key;
      currentArray = [];
      continue;
    }

    // Scalar: quoted string
    if (
      (value.startsWith('"') && value.endsWith('"')) ||
      (value.startsWith("'") && value.endsWith("'"))
    ) {
      result[key] = value.slice(1, -1);
      currentKey = key;
      continue;
    }

    // Scalar: boolean
    if (value === "true") {
      result[key] = true;
      currentKey = key;
      continue;
    }
    if (value === "false") {
      result[key] = false;
      currentKey = key;
      continue;
    }

    // Scalar: number
    if (/^-?\d+(\.\d+)?$/.test(value)) {
      result[key] = Number(value);
      currentKey = key;
      continue;
    }

    // Scalar: null
    if (value === "null" || value === "~") {
      result[key] = null;
      currentKey = key;
      continue;
    }

    // Scalar: plain string
    result[key] = value;
    currentKey = key;
  }

  // Flush any trailing array
  if (currentArray && currentKey) {
    result[currentKey] = currentArray;
  }

  return result;
}

// ─── YAML Serialization ──────────────────────────────────────────────────────

/**
 * Serialize a record to simple YAML frontmatter string.
 * Produces clean YAML for agent frontmatter blocks.
 */
export function serializeYaml(data: Record<string, unknown>): string {
  const lines: string[] = [];

  for (const [key, value] of Object.entries(data)) {
    if (value === undefined || value === null) continue;

    if (Array.isArray(value)) {
      if (value.length === 0) continue;
      lines.push(`${key}:`);
      for (const item of value) {
        lines.push(`  - ${item}`);
      }
    } else if (typeof value === "string") {
      // Quote strings that contain special YAML characters
      if (
        value.includes(":") ||
        value.includes("#") ||
        value.includes("{") ||
        value.includes("[") ||
        value.startsWith(" ") ||
        value.endsWith(" ")
      ) {
        lines.push(`${key}: "${value.replace(/"/g, '\\"')}"`);
      } else {
        lines.push(`${key}: ${value}`);
      }
    } else if (typeof value === "boolean" || typeof value === "number") {
      lines.push(`${key}: ${value}`);
    }
  }

  return lines.join("\n");
}

/**
 * Wrap content with YAML frontmatter.
 */
export function withFrontmatter(
  frontmatter: Record<string, unknown>,
  body: string,
): string {
  const yaml = serializeYaml(frontmatter);
  if (!yaml) return body;
  return `---\n${yaml}\n---\n\n${body}`;
}

// ─── File Discovery ──────────────────────────────────────────────────────────

/**
 * Read all markdown files from a directory and parse them as canonical agents.
 */
export async function readCanonicalAgents(
  agentsDir: string,
): Promise<CanonicalAgent[]> {
  const agents: CanonicalAgent[] = [];
  const glob = new Bun.Glob("*.md");

  try {
    for await (const entry of glob.scan({ cwd: agentsDir })) {
      if (entry === ".gitkeep") continue;
      const raw = await Bun.file(join(agentsDir, entry)).text();
      const { frontmatter, body } = parseFrontmatter(raw);
      const name = entry.replace(/\.md$/, "");
      agents.push({ name, body, frontmatter });
    }
  } catch {
    // agentsDir doesn't exist
  }

  return agents;
}

/**
 * Read brain.config.json from the project root.
 */
export async function readBrainConfig(
  projectRoot: string,
): Promise<BrainConfig> {
  const configFile = Bun.file(join(projectRoot, "brain.config.json"));
  return configFile.json() as Promise<BrainConfig>;
}

/**
 * Recursively collect all files in a directory, returning relative paths.
 * Uses Bun.Glob for recursive traversal.
 */
export async function collectFiles(
  dir: string,
): Promise<string[]> {
  const result: string[] = [];
  const glob = new Bun.Glob("**/*");

  try {
    for await (const entry of glob.scan({ cwd: dir })) {
      if (entry.includes("node_modules") || entry.includes(".git") || entry.includes(".DS_Store"))
        continue;
      result.push(entry);
    }
  } catch {
    // dir doesn't exist
  }

  return result;
}

// ─── Naming ──────────────────────────────────────────────────────────────────

/** Add the Brain emoji prefix to a filename. */
export function brainPrefix(name: string): string {
  if (name.startsWith("\u{1F9E0}-")) return name;
  return `\u{1F9E0}-${name}`;
}

File: apps/tui/internal/adapters/shared.go
// Package adapters provides cross-platform plugin transforms for Brain.
//
// Each adapter converts canonical content (agents, skills, commands, protocols,
// hooks, MCP) into a target platform's plugin format. Shared types and helpers
// live here; platform-specific logic lives in cursor.go and claudecode.go.
package adapters

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

// ─── Types ──────────────────────────────────────────────────────────────────

// GeneratedFile represents a single file produced by an adapter transform.
type GeneratedFile struct {
	// RelativePath is the path relative to the target platform's config directory.
	RelativePath string
	// Content is the file's text content.
	Content string
}

// BrainConfig represents the top-level brain.config.json structure.
type BrainConfig struct {
	Version   string                       `json:"version"`
	Targets   map[string]json.RawMessage   `json:"targets"`
	Agents    map[string]AgentToolConfig   `json:"agents"`
	Hooks     map[string]json.RawMessage   `json:"hooks"`
	Skills    map[string]json.RawMessage   `json:"skills"`
	Commands  map[string]json.RawMessage   `json:"commands"`
	Protocols map[string]json.RawMessage   `json:"protocols"`
}

// AgentToolConfig maps platform names to their agent config (or null).
// The "source" key holds the canonical agent path; platform keys hold
// AgentPlatformConfig or null (meaning "not for this platform").
type AgentToolConfig map[string]json.RawMessage

// AgentPlatformConfig holds per-platform agent frontmatter fields.
type AgentPlatformConfig struct {
	Model        string   `json:"model,omitempty"`
	AllowedTools []string `json:"tools,omitempty"`
	Memory       string   `json:"memory,omitempty"`
	Color        string   `json:"color,omitempty"`
	Description  string   `json:"description,omitempty"`
	ArgumentHint string   `json:"argument_hint,omitempty"`
	Skills       []string `json:"skills,omitempty"`
}

// HookPlatformConfig holds per-platform hook configuration.
type HookPlatformConfig struct {
	Event   string `json:"event"`
	Matcher string `json:"matcher,omitempty"`
	Timeout int    `json:"timeout,omitempty"`
}

// CanonicalAgent represents a parsed canonical agent markdown file.
type CanonicalAgent struct {
	// Name is the filename without extension (e.g., "architect").
	Name string
	// Body is the raw markdown body (everything after frontmatter).
	Body string
	// Frontmatter contains any YAML frontmatter fields from the canonical file.
	Frontmatter map[string]any
}

// JsonMergePayload is used for additive config merging in Cursor.
// The Go CLI reads these and merges them into the user's existing config
// without overwriting non-Brain keys.
type JsonMergePayload struct {
	ManagedKeys []string               `json:"managedKeys"`
	Content     map[string]any `json:"content"`
}

// ─── Brain Prefix ───────────────────────────────────────────────────────────

const brainEmoji = "\U0001F9E0"

// BrainPrefix adds the Brain emoji prefix to a filename if not already present.
func BrainPrefix(name string) string {
	prefix := brainEmoji + "-"
	if strings.HasPrefix(name, prefix) {
		return name
	}
	return prefix + name
}

// ─── Frontmatter Parsing ────────────────────────────────────────────────────

var frontmatterRE = regexp.MustCompile(`(?s)^---\n(.*?)\n---\n?(.*)$`)

// ParseFrontmatter splits a markdown file into its YAML frontmatter and body.
// Returns empty frontmatter if none found.
func ParseFrontmatter(raw string) (frontmatter map[string]any, body string) {
	match := frontmatterRE.FindStringSubmatch(raw)
	if match == nil {
		return map[string]any{}, raw
	}
	fm := ParseSimpleYAML(match[1])
	return fm, strings.TrimLeft(match[2], "\n")
}

// ParseSimpleYAML is a minimal YAML parser for agent frontmatter.
// Handles: scalars, simple arrays (inline [...] and block - item), one-level nesting.
// Does NOT handle complex nested structures, multi-line strings, or anchors.
func ParseSimpleYAML(yaml string) map[string]any {
	result := make(map[string]any)
	lines := strings.Split(yaml, "\n")
	var currentKey string
	var currentArray []string

	for _, line := range lines {
		trimmed := strings.TrimSpace(line)

		// Skip empty lines and comments
		if trimmed == "" || strings.HasPrefix(trimmed, "#") {
			if currentArray != nil && currentKey != "" {
				result[currentKey] = toInterfaceSlice(currentArray)
				currentArray = nil
				currentKey = ""
			}
			continue
		}

		// Block array item (  - value)
		if currentArray != nil && strings.HasPrefix(trimmed, "- ") {
			currentArray = append(currentArray, strings.TrimSpace(strings.TrimPrefix(trimmed, "- ")))
			continue
		}

		// Flush pending array
		if currentArray != nil && currentKey != "" {
			result[currentKey] = toInterfaceSlice(currentArray)
			currentArray = nil
		}

		// Key-value pair
		key, rawValue, found := strings.Cut(trimmed, ":")
		if !found {
			continue
		}
		rawValue = strings.TrimSpace(rawValue)

		// Inline array: [a, b, c]
		if strings.HasPrefix(rawValue, "[") && strings.HasSuffix(rawValue, "]") {
			inner := rawValue[1 : len(rawValue)-1]
			items := strings.Split(inner, ",")
			var cleaned []string
			for _, item := range items {
				s := strings.TrimSpace(item)
				if s != "" {
					cleaned = append(cleaned, s)
				}
			}
			result[key] = toInterfaceSlice(cleaned)
			currentKey = key
			continue
		}

		// Empty value followed by array items
		if rawValue == "" {
			currentKey = key
			currentArray = []string{}
			continue
		}

		// Scalar: quoted string
		if (strings.HasPrefix(rawValue, `"`) && strings.HasSuffix(rawValue, `"`)) ||
			(strings.HasPrefix(rawValue, "'") && strings.HasSuffix(rawValue, "'")) {
			result[key] = rawValue[1 : len(rawValue)-1]
			currentKey = key
			continue
		}

		// Scalar: boolean
		if rawValue == "true" {
			result[key] = true
			currentKey = key
			continue
		}
		if rawValue == "false" {
			result[key] = false
			currentKey = key
			continue
		}

		// Scalar: null
		if rawValue == "null" || rawValue == "~" {
			result[key] = nil
			currentKey = key
			continue
		}

		// Scalar: plain string
		result[key] = rawValue
		currentKey = key
	}

	// Flush trailing array
	if currentArray != nil && currentKey != "" {
		result[currentKey] = toInterfaceSlice(currentArray)
	}

	return result
}

func toInterfaceSlice(ss []string) []any {
	out := make([]any, len(ss))
	for i, s := range ss {
		out[i] = s
	}
	return out
}

// ─── YAML Serialization ─────────────────────────────────────────────────────

// SerializeYAML converts a map to a simple YAML frontmatter string.
func SerializeYAML(data map[string]any) string {
	var lines []string

	for key, value := range data {
		if value == nil {
			continue
		}

		switch v := value.(type) {
		case []any:
			if len(v) == 0 {
				continue
			}
			lines = append(lines, key+":")
			for _, item := range v {
				lines = append(lines, fmt.Sprintf("  - %v", item))
			}
		case []string:
			if len(v) == 0 {
				continue
			}
			lines = append(lines, key+":")
			for _, item := range v {
				lines = append(lines, fmt.Sprintf("  - %s", item))
			}
		case string:
			if strings.ContainsAny(v, ":#{}[]") || strings.HasPrefix(v, " ") || strings.HasSuffix(v, " ") {
				lines = append(lines, fmt.Sprintf(`%s: "%s"`, key, strings.ReplaceAll(v, `"`, `\"`)))
			} else {
				lines = append(lines, fmt.Sprintf("%s: %s", key, v))
			}
		case bool:
			lines = append(lines, fmt.Sprintf("%s: %t", key, v))
		case int:
			lines = append(lines, fmt.Sprintf("%s: %d", key, v))
		case float64:
			lines = append(lines, fmt.Sprintf("%s: %g", key, v))
		}
	}

	return strings.Join(lines, "\n")
}

// WithFrontmatter wraps content with YAML frontmatter.
func WithFrontmatter(frontmatter map[string]any, body string) string {
	yaml := SerializeYAML(frontmatter)
	if yaml == "" {
		return body
	}
	return "---\n" + yaml + "\n---\n\n" + body
}

// ─── File Discovery ─────────────────────────────────────────────────────────

// ReadCanonicalAgents reads all markdown files from a directory and parses
// them as canonical agents.
func ReadCanonicalAgents(agentsDir string) ([]CanonicalAgent, error) {
	entries, err := os.ReadDir(agentsDir)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, fmt.Errorf("read agents dir: %w", err)
	}

	var agents []CanonicalAgent
	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".md") {
			continue
		}
		if entry.Name() == ".gitkeep" {
			continue
		}

		raw, err := os.ReadFile(filepath.Join(agentsDir, entry.Name()))
		if err != nil {
			return nil, fmt.Errorf("read agent %s: %w", entry.Name(), err)
		}

		fm, body := ParseFrontmatter(string(raw))
		name := strings.TrimSuffix(entry.Name(), ".md")
		agents = append(agents, CanonicalAgent{
			Name:        name,
			Body:        body,
			Frontmatter: fm,
		})
	}

	return agents, nil
}

// ReadBrainConfig reads brain.config.json from the project root.
func ReadBrainConfig(projectRoot string) (*BrainConfig, error) {
	data, err := os.ReadFile(filepath.Join(projectRoot, "brain.config.json"))
	if err != nil {
		return nil, fmt.Errorf("read brain.config.json: %w", err)
	}

	var config BrainConfig
	if err := json.Unmarshal(data, &config); err != nil {
		return nil, fmt.Errorf("parse brain.config.json: %w", err)
	}

	return &config, nil
}

// GetAgentPlatformConfig extracts the platform-specific config for an agent.
// Returns nil if the agent is not configured for this platform, or if the
// platform key is explicitly set to null.
func GetAgentPlatformConfig(agentConfig AgentToolConfig, platform string) *AgentPlatformConfig {
	raw, ok := agentConfig[platform]
	if !ok {
		return nil
	}

	// Check for explicit null
	if string(raw) == "null" {
		return nil
	}

	var config AgentPlatformConfig
	if err := json.Unmarshal(raw, &config); err != nil {
		return nil
	}

	return &config
}

// ScanMarkdownFiles returns all .md filenames (without path) in a directory.
func ScanMarkdownFiles(dir string) ([]string, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}

	var files []string
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		if entry.Name() == ".gitkeep" || entry.Name() == ".DS_Store" {
			continue
		}
		if strings.HasSuffix(entry.Name(), ".md") {
			files = append(files, entry.Name())
		}
	}

	return files, nil
}

// ScanAllFiles returns all non-hidden filenames in a directory.
func ScanAllFiles(dir string) ([]string, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}

	var files []string
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		name := entry.Name()
		if name == ".gitkeep" || name == ".DS_Store" || strings.HasPrefix(name, ".") {
			continue
		}
		files = append(files, name)
	}

	return files, nil
}

// WriteGeneratedFiles writes a slice of generated files to the given output directory.
// Creates parent directories as needed.
func WriteGeneratedFiles(files []GeneratedFile, outputDir string) error {
	for _, f := range files {
		fullPath := filepath.Join(outputDir, f.RelativePath)
		if err := os.MkdirAll(filepath.Dir(fullPath), 0755); err != nil {
			return fmt.Errorf("create dir for %s: %w", f.RelativePath, err)
		}
		if err := os.WriteFile(fullPath, []byte(f.Content), 0644); err != nil {
			return fmt.Errorf("write %s: %w", f.RelativePath, err)
		}
	}
	return nil
}

// CollectFiles recursively collects all files in a directory, returning
// relative paths. Skips node_modules, .git, and .DS_Store.
func CollectFiles(dir string) ([]string, error) {
	var result []string

	err := filepath.WalkDir(dir, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return err
		}

		name := d.Name()
		if d.IsDir() {
			if name == "node_modules" || name == ".git" {
				return filepath.SkipDir
			}
			return nil
		}

		if name == ".DS_Store" {
			return nil
		}

		rel, err := filepath.Rel(dir, path)
		if err != nil {
			return err
		}
		result = append(result, rel)
		return nil
	})
	if err != nil && !os.IsNotExist(err) {
		return nil, err
	}

	return result, nil
}
